src/core/Target_location.ml

```diff
@@ -0,0 +1,82 @@
+(* Cooper Pierce
+ *
+ * Copyright (C) 2023 Semgrep Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+type t = Code of code | Lockfile of lockfile [@@deriving show]
+
+(* The location of a "normal" semgrep target, comprising source code (or for
+ * regex/generic, arbitrary text data) to be executed.
+ *)
+and code = {
+  (* The source of the data as is relevant to the user. This could be, e.g., a
+   * relative (from the project root or scanning directory) path to a file, a
+   * git object and associated information, or anything else a Source.t can
+   * designate.
+   *
+   * This should be used when reporting a location to the user.
+   *)
+  source : Source.t;
+  (* The path to a file which contains the data to be scanned. This could be
+   * the same as the source, if the source is a path to a regular file, or it
+   * could be a tempfile
+   *
+   *)
+  file : Fpath.t;
+  analyzer : Xlang.t;
+  products : Semgrep_output_v1_t.product list;
+  (* Optional lockfile associated with this target.
+   * The association is namely that this target has its dependencies specified
+   * by the given lockfile. Core doesn't need to worry about determining these
+   * associations; rather the target selection and generation process must
+   * resolve these connections as part of generating code targets.
+   *)
+  lockfile : lockfile option;
+}
+[@@deriving show]
+
+(* A lockfile to be matched against dependency patterns. *)
+and lockfile = {
+  source : Source.t;
+  file : Fpath.t;
+  kind : Lockfile_kind.t;
+  manifest : manifest option;
+}
+[@@deriving show]
+
+and manifest = { source : Source.t; file : Fpath.t; kind : Manifest_kind.t }
+[@@deriving show]
```

Comment:
These are explained in `Lockfile_target.ml`, perhaps the comments should be move or duplicated.

Reply:
> I would prefer [...] that the fields be unified.

I did consider defining this and `Lockfile_target` as

```ocaml
type 'content t = {
  source : Source.t;
  file : Fpath.t;
  kind : Lockfile_kind.t;
  manifest : manifest_target option;
  content: 'content;
}
type target_content = {
  lazy_content : string lazy_t;
  lazy_ast_and_errors : Dependency.t list lazy_t;
}
type location = unit t;
type target = target_content t;
```
(and then similar for `code` in `Xtarget` or a substantially similar re-named module)

but this is at odds with having a unified `Target_location.t` type which shares fields, rather than being a sum of `Xtarget.location` and `Lockfile_target.location` as defined above.

Given that, which would you prefer?

Reply:
I still prefer the unified `Target_location.t`. My goal is for people changing fields for something specific to their product to not think about the generic `Target_location`, and for people changing something generic about the `Target_location` to not think about the product-specific fields.

Reply:
Re fields: Addressed in fc35371b4020d3fdf1beb37d3446dbe5f1e81f70. Re explanation: some commit added docs; see full diff.